app/controllers/maps_controller.rb

```diff
@@ -21,9 +21,49 @@ def create
     end
   end
 
+  def edit
+    respond_to :js
+
+    map = current_user.maps.find(params[:id])
+    render layout: "modal", locals: { map: map }
+  end
+
+  def update
+    map = current_user.maps.find(params[:id])
+    result = UpdateMap.call(
+      campaign: map.campaign,
+      map: map,
+      map_params: map_params
+    )
+    if result.success?
+      head :ok
+    else
+      render partial: "form", locals: { map: map }, status: :bad_request
+    end
+  end
+
+  def destroy
+    map = current_user.maps.find(params[:id])
+    result = DeleteMap.call(campaign: map.campaign, map: map)
+    if result.success?
+      head :ok
+    else
+      head :bad_request
+    end
+  end
+
   private
 
   def map_params
     params.require(:map).permit(:name, :image, :grid_size)
   end
+
+  def map_model_for_form(map)
+    if map.persisted?
+      map
+    else
+      [map.campaign, map]
+    end
+  end
+  helper_method :map_model_for_form
```

Comment:
Seeing as we never link to maps directly, we could get rid of the `:shallow` nested resource so we don't have to worry about the appropriate model?

Reply:
Ok, I've done this in 0f0f900.